src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -310,6 +289,70 @@ AcquireExecutorShardLock(Task *task, LOCKMODE lockMode)
 
 
 /*
+ * AcquireExecutorMultiShardLocks acquires shard locks need for execution
+ * of writes on multiple shards.
+ *
+ * 1. If citus.all_modifications_commutative is set to true, then all locks
+ * are acquired as ShareUpdateExclusiveLock.
+ * 2. If citus.all_modifications_commutative is false, then only the shards
+ * with 2 or more replicas are locked with ExclusiveLock. Otherwise, the
+ * lock is acquired with ShareUpdateExclusiveLock.
+ *
+ * ShareUpdateExclusiveLock conflicts with itself such that only one
```

Comment:
I don't really want to lock first_table for modification, since this is probably the raw events tables that is constantly taking writes, but it could cause second_table's shard placement to diverge. Roll-ups using INSERT/SELECT should typically converge by only selecting items that have not yet been processed to prevent duplication, though it'd be nicer if we can actually guarantee this.

The crude solutions are to prevent INSERT/SELECT for replication factor > 1 or block writes on the shards used in the SELECT. In principle, we only need to block modifications for as long as it takes the remote commands on all placements to obtain a snapshot. A more efficient solution could be to set the remote transaction isolation level to repeatable read, block writes on the shards from which we SELECT, perform a trivial SELECT query to force the remote transaction to obtain a snapshot, unblock writes, and continue executing the insert/select command.